### PR TITLE
Fix EigenProblem scaling to work with source terms

### PR DESCRIPTION
**Pull request by loganharbour**
Refs #16225

### CODE REVIEW

**Comment by loganharbour** [comment]
I'm working on a test that will be akin to DSA for transport.

**Comment by YaqiWang** [comment]
Normalization for a source problem is not valid. I am away from my laptop and will get to this ASAP.

File: framework/src/problems/EigenProblem.C [hunk 1]
```diff
@@ -517,7 +523,11 @@ EigenProblem::solve()
       {
         _console << " Free power iteration starts" << std::endl;
         doFreeNonlinearPowerIterations(solverParams()._free_power_iterations);
-        _first_solve = false;
+
+        // If there is a source term in the nonlinear problem, the scaling
+        // on the initial condition for ||Bx||_2 = inverse of the eigenvalue
+        // is no longer valid so we will need to re-scale again
+        preScaleEigenVector();
```
Comment (YaqiWang) [inline framework/src/problems/EigenProblem.C:530]: Power iteration is solving Ax = Bx^old / ||Bx^old||_2, and k = ||Bx||_2, so `preScaleEigenVector()` will be a null operation here. Code wise, `_nl_eigen->getConvergedEigenvalue(_active_eigen_index)` will just return ||Bx||_2, so the scaling factor will be one.

File: framework/src/problems/EigenProblem.C [hunk 2]
```diff
@@ -504,8 +505,13 @@ EigenProblem::solve()
...
     _nl_eigen->attachSLEPcCallbacks();
 
-    // Scale eigen vector if necessary
-    preScaleEigenVector();
+    // We need to make sure that ||Bx||_2 is equal to inverse of eigenvalue for
+    // the NEWTON iteration iteration
+    // - On the first solve, we will scale so that this is true for the initial guess
+    // - If the user has a custom normalization, we will also need to re-scale
+    //   back to the required normalization for the NEWTON solve
+    if (_first_solve || _has_normalization)
```
Comment (YaqiWang) [inline framework/src/problems/EigenProblem.C:513]: This is good if the solution x is not altered outside by users when reentering. But to be safe, I did not have check to make sure we enter the solver properly, i.e. ||Bx||_2 is equal to current eigenvalue. This is debatable. I think it is fine to have this. But we need to add a comment above that `even we do not have this condition, the solver can tolerate it`.

File: framework/src/problems/EigenProblem.C [hunk 3]
```diff
@@ -540,8 +550,11 @@ EigenProblem::solve()
     _nl->solve();
     _nl->update();
 
-    // Scale eigen vector if users ask
-    postScaleEigenVector();
+    _first_solve = false;
```
Comment (YaqiWang) [inline framework/src/problems/EigenProblem.C:553]: Doing this move is not necessary though. `_first_solve` is used only for initial free power iterations. Moving it to here makes things not so obvious.

File: framework/src/problems/EigenProblem.C [hunk 4]
```diff
@@ -409,52 +409,53 @@ EigenProblem::preScaleEigenVector()
...
+  while (!MooseUtils::absoluteFuzzyEqual(v, c))
+  {
+    // If postprocessor is not defined on eigen variables, scaling might not work
+    if (itr > 10)
+      mooseError("Can not scale eigenvector to the required factor ",
+                 v,
+                 " please check if postprocessor is defined on only eigen variables");
 
-      mooseAssert(c != 0., "postprocessor value used for scaling can not be zero");
+    if (c == 0)
```
Comment (YaqiWang) [inline framework/src/problems/EigenProblem.C:448]: 👍

File: framework/src/problems/EigenProblem.C [hunk 5]
```diff
@@ -540,8 +550,11 @@ EigenProblem::solve()
     _nl->solve();
     _nl->update();
 
-    // Scale eigen vector if users ask
-    postScaleEigenVector();
+    _first_solve = false;
+
+    // Scale the eigen vector ONLY if the user has a custom normalization
+    if (_has_normalization)
```
Comment (YaqiWang) [inline framework/src/problems/EigenProblem.C:556]: This change is fine.

**Review by YaqiWang — CHANGES_REQUESTED** [review]
I do not think this solves the problem of having normalization for DSA. When we have DSA, we should not have normalization. There are a few good polish in the attempt that we can keep, but the issue is invalid.

File: framework/src/problems/EigenProblem.C (hunk 3 above)
Comment (fdkong) [reply framework/src/problems/EigenProblem.C:553]: Motivation is that: if you not have "free_power", and then `_first_solve` will be always true. I like this change

Comment (YaqiWang) [reply framework/src/problems/EigenProblem.C:553]: So you want `_first_solve` used somewhere else? Now I am thinking that we can wrap the free power iterations with `MooseDoOnce` then this flag is not visible.

File: framework/src/problems/EigenProblem.C (hunk 1 above)
Comment (YaqiWang) [reply framework/src/problems/EigenProblem.C:530]: Like
```
      if (solverParams()._free_power_iterations)
      {
       MooseDoOnce(
        _console << " Free power iteration starts" << std::endl;
        doFreeNonlinearPowerIterations(solverParams()._free_power_iterations);
        );
      }
```

File: framework/src/problems/EigenProblem.C (hunk 3 above)
Comment (YaqiWang) [reply framework/src/problems/EigenProblem.C:553]: Like
```
      if (solverParams()._free_power_iterations)
      {
       MooseDoOnce(
        _console << " Free power iteration starts" << std::endl;
        doFreeNonlinearPowerIterations(solverParams()._free_power_iterations);
        );
      }
```

**Pull request closed without merging**